Fill.py

```diff
@@ -62,18 +62,137 @@ def fill_restrictive(multiworld: MultiWorld, base_state: CollectionState, locati
+                batch_size_float = max_batch_size * average_remaining / largest_remaining
+            else:
+                batch_size_float = max_batch_size
+
+            # Round to the nearest integer.
+            batch_size = round(batch_size_float)
+            # Limit the minimum number of items per player in the batch.
+            batch_size = max(min_batch_size, batch_size)
+            # Don't make the batch larger than the largest remaining item pool.
+            batch_size = min(batch_size, largest_remaining)
+
+            # Set per-batch variables.
+
+            batch_per_player_placements_remaining = batch_size
+
+            # If a player has fewer items remaining than the size of the batch, then that player has some empty spaces
+            # for items in the batch. This allows for items belonging to that player, that were displaced by a swap, to
+            # be added to the current batch until the spaces are used up.
+            batch_empty_spaces = {player: batch_size - min(len(player_items), batch_size)
+                                  for player, player_items in reachable_items.items()}
+
+            # Iterate the first `batch_size` items of each player's remaining items into a list of all items that will
+            # be placed in the current batch.
+            # Items to place are picked from the end of each player's item pool, so, to get the items in the order they
+            # will be placed, the item pools must be iterated in reverse.
+            item_iters = [reversed(items) for items in reachable_items.values() if items]
+            batch_item_pool = [item for item_iter in item_iters
+                               for item in itertools.islice(item_iter, batch_size)]
+
+            # Collect the remaining items, which won't be placed in this batch, into a copy of `base_state` and make
+            # that the base state for the batch.
+            batch_base_state = base_state.copy()
+            for item_iter in item_iters:
+                for item in item_iter:
+                    batch_base_state.collect(item, True)
+
+            # If there was an existing batch sweep state, it must be re-created because it will have collected items
+            # that are being placed in the current batch. The placement loop is responsible for creating the sweep state
+            # because the sweep state may get destroyed by a swap.
+            batch_sweep_state = None
+
         # grab one item per player
         items_to_place = [items.pop()
                           for items in reachable_items.values() if items]
         for item in items_to_place:
+            for p, batch_pool_item in enumerate(batch_item_pool):
+                if batch_pool_item is item:
+                    batch_item_pool.pop(p)
+                    break
             for p, pool_item in enumerate(item_pool):
                 if pool_item is item:
                     item_pool.pop(p)
                     break
```

Comment:
I think there's a reason for `item_pool` to be done this way where `batch_item_pool` wouldn't have the same reason.

`item_pool` is an object passed as a parameter to this function that is mutated. Other code might depend on the pointer not changing which object it's pointing to.

`batch_item_pool` is not passed as a parameter to this function, so it's not important to keep the pointer pointing at the same object.
So we could gather up the indexes that need to be removed and then replace `batch_item_pool` with a new list with the gathered indexes filtered out, reducing the time complexity by one exponent.

In my trace, the `batch_item_pool.pop(p)` seemed to always pop index 0, which is the slowest `pop`.
`item_pool.pop(p)` was popping from the end, which is the fastest `pop`.

I didn't measure the performance of the suggested change, so I'm not sure whether it's worth it or even bad.

Reply:
I determined it to be irrelevant to performance. I also suspect you ran with only a single player in your generation. I think that if you were to have more than one player, all the items for each player would be added into `batch_item_pool` in a row, so the indices to pop for players after the first would not be at the end of the list.

`item_pool` does not need to be done this way either, it's clearly more efficient to iterate in reverse, but it's even more efficient to replace the entire `item_pool` at the start with a dictionary keyed by the `id()` of each item. But both of these changes are basically irrelevant unless you're generating with many many worlds with both fast access rules and a large percentage of progression items. e.g. generating with 400 default AHiT yamls (~43200 progression items out of ~89974 items total) goes from about 90s to 50s on main (with progression balancing disabled and --skip_output).